[Temporal] Add PlainYearMonth `from` method

catamorphism opened the pull request:
#### 7af7345ce5192ca27a5461ec5222abea88f24443
<pre>
[Temporal] Add PlainYearMonth `from` method
<a href="https://bugs.webkit.org/show_bug.cgi?id=303503">https://bugs.webkit.org/show_bug.cgi?id=303503</a>

Reviewed by Darin Adler and Sosuke Suzuki.

Add this method.

Co-authored-by: Darin Adler &lt;33187212+darinadler@users.noreply.github.com&gt;

* JSTests/stress/temporal-plainyearmonth.js:
(shouldThrow):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::isoDateFromFields):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
(JSC::TemporalPlainYearMonth::from):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h:
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/304135@main">https://commits.webkit.org/304135@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e13a170ec90401c33ab7f0044c8433d31b1cd68

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142165 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86585 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7698 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6951 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102908 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70190 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137587 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5402 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120681 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83711 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5258 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2869 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2756 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126688 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38822 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144857 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133148 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6772 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39403 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6846 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5679 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111596 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28306 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5091 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116955 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60657 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6823 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35147 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166059 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6629 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70405 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43403 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6865 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6738 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ptomato commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,105 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+            RETURN_IF_EXCEPTION(scope, { });
+        }
+        RELEASE_AND_RETURN(scope, { });
+    }
```
This may be dead code, I think it's already covered in the earlier `if (isString)` block?

catamorphism replied:
You're right; fixed in [1a4c1ed](https://github.com/WebKit/WebKit/pull/54800/commits/1a4c1ed1a6b5d1c68ff7e88c90b9c01a54096bcd)

ptomato commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,105 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+            return jsCast<TemporalPlainYearMonth*>(itemValue);
+
+        if (itemValue.inherits<TemporalPlainYearMonth>())
+            return TemporalPlainYearMonth::create(vm, globalObject->plainYearMonthStructure(), jsCast<TemporalPlainYearMonth*>(itemValue)->plainYearMonth());
```
This may also be dead code, but anyway it should only create a copy of the object if it's called from `Temporal.PlainYearMonth.from()` in userland; otherwise it should return the same object.

catamorphism replied:
You're right; fixed in [1a4c1ed](https://github.com/WebKit/WebKit/pull/54800/commits/1a4c1ed1a6b5d1c68ff7e88c90b9c01a54096bcd)

darinadler commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,105 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+    auto dateTime = ISO8601::parseCalendarDateTime(string, TemporalDateFormat::YearMonth);
+    if (dateTime) [[likely]] {
+        auto [plainDate, plainTimeOptional, timeZoneOptional, calendarOptional] = WTFMove(dateTime.value());
+        if (calendarOptional && StringView(calendarOptional.value()) != String::fromLatin1("iso8601")) [[unlikely]] {
```
Please do not use `String::fromLatin1` for ASCII literals. We need to remove this from existing code and stop adding it in new code!
```suggestion
        if (calendarOptional && calendarOptional.value() != "iso8601"_s) [[unlikely]] {
```

catamorphism replied:
Happy to take out the `fromLatin1`, but it won't compile without the `StringView` (`calendarOptional.value()` is a `Vector<Latin1Character, maxCalendarLength>`).

darinadler replied:
Best idiom for comparing a `Vector` of `Latin1Character` with a string constant is to call the `equal` function that compares two spans, something like:
```suggestion
        if (calendarOptional && !equal(calendarOptional->span(), "iso8601"_span8)) [[unlikely]] {
```
Using `StringView` brings branches for handling of 16-bit characters into play, unnecessarily.

catamorphism replied:
Done.

darinadler commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h:
```diff
@@ -47,6 +47,9 @@ class TemporalPlainYearMonth final : public JSNonFinalObject {
 
     DECLARE_INFO;
 
+    static TemporalPlainYearMonth* from(JSGlobalObject*, JSValue, std::optional<JSValue>);
```
I don’t see any code passing `nullopt` for this third argument. Are you sure that code path is exercised?

`JSValue` has a distinct zero value that is distinct from JavaScript null and undefined, so it can be used for a `JSValue` that might be omitted, passing zero value as appropriate. And often in other JavaScript contexts we can use null and undefined instead. Do we really need to use `std::optional` here? I can’t find where it is used and for what.

Typically when called from JavaScript, omitted arguments are represented with an `JSValue` of undefined.

Also, the argument meaning isn’t clear here without an argument name, so we should include that name here.

catamorphism replied:
I think the reason it was optional is that it's used that way in a future commit (I originally had a big monolithic PR for these changes and what I'm submitting now is the result of splitting it up), but using the zero value for `JSValue` seems fine.

darinadler replied:
I suspect we might want to use undefined rather than the zero value; I’d probably have to see more of the future code to know if that’s a good choice. But I also would have probably suggested making it non-optional and then add the ability to leave it out in that future commit.

darinadler commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.cpp:
```diff
@@ -35,6 +35,7 @@ namespace JSC {
 
 STATIC_ASSERT_IS_TRIVIALLY_DESTRUCTIBLE(TemporalPlainYearMonthConstructor);
 
+static JSC_DECLARE_HOST_FUNCTION(temporalPlainYearMonthConstructorFuncFrom);
```
```suggestion
static JSC_DECLARE_HOST_FUNCTION(temporalPlainYearMonthConstructorFuncFrom);

```

darinadler commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h:
```diff
@@ -47,6 +47,9 @@ class TemporalPlainYearMonth final : public JSNonFinalObject {
     DECLARE_INFO;
 
+    static TemporalPlainYearMonth* from(JSGlobalObject*, JSValue, std::optional<JSValue>);
+    static TemporalPlainYearMonth* from(JSGlobalObject*, WTF::String);
```
This should be `StringView`, not `String`. Also, we almost never should need to write `WTF::String` instead of `String`.

catamorphism replied:
Fixed in [4c22618](https://github.com/WebKit/WebKit/pull/54800/commits/4c2261859a582283eb401ea25a5e1616dfac6080)

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,91 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+        // See step 11 of ToTemporalYearMonth
+        if (optionsValue)
+            toTemporalOverflow(globalObject, optionsValue.value());
+        RETURN_IF_EXCEPTION(scope, { });
```
Should be like this?:

```cpp
        if (optionsValue) {
            toTemporalOverflow(globalObject, optionsValue.value());
            RETURN_IF_EXCEPTION(scope, { });
        }
```

catamorphism replied:
Fixed in [4c22618](https://github.com/WebKit/WebKit/pull/54800/commits/4c2261859a582283eb401ea25a5e1616dfac6080)

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,91 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+        RELEASE_AND_RETURN(scope, result);
+    }
+
+    std::optional<JSObject*> options;
```
I think we should use `JSObject*` directly instead of `std::optional` and `nullptr` instead of `nullopt`, following the　other JSC code.

catamorphism replied:
Changed in [4c22618](https://github.com/WebKit/WebKit/pull/54800/commits/4c2261859a582283eb401ea25a5e1616dfac6080)

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,91 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+    }
+
+    throwRangeError(globalObject, scope,
+        makeString("Temporal.PlainYearMonth.from: invalid date string "_s, string));
```
I think we should use `tryMakeString` instead of `makeString` like https://github.com/WebKit/WebKit/blob/21ec19ce100f4f11702815b94c1f9f1faa59b791/Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp#L746-L751

catamorphism replied:
Changed in [4c22618](https://github.com/WebKit/WebKit/pull/54800/commits/4c2261859a582283eb401ea25a5e1616dfac6080)

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,94 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+    }
+
+    JSObject* options = nullptr;
+    if (optionsValue) {
```
`optionsValue` is now just `JSValue`, so `if (optionsValue)` means checking if `optionsValue` is not zero value.
But I think what we should do here is checking if `optionsValue` is not `undefined` (or `null`) ?

sosukesuzuki replied:
If my comment makes sense, we should add more tests to check this behavior

catamorphism replied:
I think you're right, checking for `undefined` makes sense, because that's what would be passed in from `temporalPlainYearMonthConstructorFuncFrom` if no options argument was provided. I changed it. I also added stress tests for null and explicit-undefined options objects.

darinadler commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,95 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+    if (!optionsValue.isUndefined()) {
+        options = intlGetOptionsObject(globalObject, optionsValue);
+        RETURN_IF_EXCEPTION(scope, { });
+    }
```
The `intlGetOptionsObject` function already handles undefined this way, returns nullptr with no exception, so the code needn’t add a special case for it.
```suggestion
    JSObject* options = intlGetOptionsObject(globalObject, optionsValue);
    RETURN_IF_EXCEPTION(scope, { });
```

darinadler commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,95 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+    if (isString) {
+        auto string = item.toWTFString(globalObject);
+        RETURN_IF_EXCEPTION(scope, { });
+        result = TemporalPlainYearMonth::from(globalObject, string);
```
The local variable `isString` doesn’t do us any good here. Also, the `JSValue` member function `getString` is designed for this kind of use and is more efficient than calling `isString` and then `toWTFString`. And `result` should be scoped inside the if statement.
```suggestion
    auto string = item.getString(globalObject);
    RETURN_IF_EXCEPTION(scope, { });
    if (!string.isNull()) {
        auto* result = TemporalPlainYearMonth::from(globalObject, string);
```

darinadler commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,95 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+    }
+
+    if (item.isObject()) {
+
```
```suggestion
```

darinadler commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -118,6 +118,95 @@ String TemporalPlainYearMonth::toString(JSGlobalObject* globalObject, JSValue op
+        if (options)
+            optionsOrOverflow = options;
+        auto overflow = TemporalOverflow::Constrain;
+        auto plainYearMonth = TemporalCalendar::isoDateFromFields(globalObject, asObject(item), TemporalDateFormat::YearMonth, optionsOrOverflow, overflow);
```
I find this a bit confusing. Is `overflow` an out argument? If so, why are we ignoring its value?

catamorphism replied:
It's an out argument, but per the spec, it isn't needed in this case (it's only used within `isoDateFromFields`). It is needed when `TemporalPlainDateTime::from` calls `isoDateFromFields`, and it's unpacked inside `isoDateFromFields` due to the order of operations that has to be followed.

darinadler replied:
I would prefer to make `isoDateFromFields` return a `pair` or `tuple` rather than using an out argument. Maybe we can do that as a later refinement.

sosukesuzuki reviewed (approved):
r=me

catamorphism committed:
[Temporal] Add PlainYearMonth `from` method
https://bugs.webkit.org/show_bug.cgi?id=303503

Reviewed by Darin Adler and Sosuke Suzuki.

Add this method.

Co-authored-by: Darin Adler <33187212+darinadler@users.noreply.github.com>

* JSTests/stress/temporal-plainyearmonth.js:
(shouldThrow):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::isoDateFromFields):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
(JSC::TemporalPlainYearMonth::from):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h:
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/304135@main
